Core/KokoroEngine.cs

```diff
@@ -8,42 +8,43 @@
 /// <remarks> Contains a background worker thread that dispatches queued jobs/actions linearly. </remarks>
 public class KokoroEngine : IDisposable {
     protected readonly KokoroModel model;
-    protected readonly ConcurrentQueue<KokoroJob> queuedJobs = [];
-
-    volatile bool hasExited;
+    protected readonly BlockingCollection<KokoroJob> queuedJobs = [];
+    readonly CancellationTokenSource cancellation = new();
 
     /// <summary> Creates a new Kokoro Engine instance, loading the model into memory and initializing a background worker thread to continuously scan for newly queued jobs, dispatching them in order, when it's free. </summary>
     /// <remarks> If 'options' is specified, the model will be loaded with them. This is particularly useful when needing to run on non-CPU backends, as the default backend otherwise is the CPU with 8 threads. </remarks>
     public KokoroEngine(string modelPath, SessionOptions options = null) {
         model = new KokoroModel(modelPath, options);
 
-        new Thread(async () => {
-            while (!hasExited) {
-                await Task.Delay(10);
-                while (!hasExited && queuedJobs.TryDequeue(out var job)) {
-                    while (!hasExited && !job.isDone) {
-                        job.Progress(model);
-                        await Task.Delay(1);
-                    }
-                }
+        new Thread(() =>
+        {
+            try {
+                while (queuedJobs.TryTake(out var kokoroJob, -1, cancellation.Token))
+                    while (!cancellation.IsCancellationRequested && !kokoroJob.isDone)
+                        kokoroJob.Progress(model);
+            }
+            catch (OperationCanceledException) {
+                // If cancellation was requested just swallow the cancellation exception and exit.
             }
+
+            foreach (var kokoroJob in queuedJobs.GetConsumingEnumerable())
+                kokoroJob.Cancel();
+            queuedJobs.Dispose();
         }).Start();
```

Comment:
No longer need the token! (+ minor style change)

```suggestion
        new Thread(() => {
            foreach (var kokoroJob in queuedJobs.GetConsumingEnumerable()) {
                while (!kokoroJob.isDone) { kokoroJob.Progress(model); }
            }
        }).Start();
```
// optional `queuedJobs.Dispose()`, but it should be empty by thread exit and properly collected anyway!

Reply:
I think you do need the token. As I understand it `CompleteAdding` just marks the collection as complete, so the enumerable will end once it has consumed **all** items. That's a change in behaviour from currently, where it will exit as soon as cancelled.

Reply:
Oh I guess you're depending on the `Dipose` call calling cancel? So the worker will pull those jobs off the queue, but will immediately exit because they've been cancelled. That makes sense.

Reply:
Yes, the original code had 2x `hasExited` checks which made it not obvious, but the inner one wasn't really needed.
Meanwhile I just love improvements like this:
![image](https://github.com/user-attachments/assets/c5172af3-b51a-47dc-bb9d-93e08081085f)